packages/react-instantsearch-core/src/components/InstantSearch.tsx

```diff
@@ -137,64 +130,63 @@ class InstantSearch extends Component<Props, State> {
     stalledSearchDelay: PropTypes.number,
   };
 
-  isControlled: boolean;
-  isUnmounting: boolean;
-  aisManager: InstantSearchManager;
-
-  constructor(props: Props) {
-    super(props);
-    this.isControlled = Boolean(props.searchState);
-    const initialState = this.isControlled ? props.searchState : {};
-    this.isUnmounting = false;
-
-    this.aisManager = createInstantSearchManager({
-      indexName: props.indexName,
-      searchClient: props.searchClient,
-      initialState,
-      resultsState: props.resultsState,
-      stalledSearchDelay: props.stalledSearchDelay,
-    });
-
-    this.state = {
+  static getDerivedStateFromProps(
+    nextProps: Props,
+    prevState: State
+  ): Partial<State> {
+    const nextIsControlled = isControlled(nextProps);
+
+    if (!prevState.isControlled && nextIsControlled) {
+      throw new Error(
+        "You can't switch <InstantSearch> from being uncontrolled to controlled"
+      );
+    }
+
+    if (prevState.isControlled && !nextIsControlled) {
+      throw new Error(
+        "You can't switch <InstantSearch> from being controlled to uncontrolled"
+      );
+    }
+
+    return {
+      isControlled: nextIsControlled,
       contextValue: {
-        onInternalStateUpdate: this.onWidgetsInternalStateUpdate.bind(this),
-        createHrefForState: this.createHrefForState.bind(this),
-        onSearchForFacetValues: this.onSearchForFacetValues.bind(this),
-        onSearchStateChange: this.onSearchStateChange.bind(this),
-        onSearchParameters: this.onSearchParameters.bind(this),
-        store: this.aisManager.store,
-        widgetsManager: this.aisManager.widgetsManager,
-        mainTargetedIndex: this.props.indexName,
+        ...prevState.contextValue,
+        mainTargetedIndex: nextProps.indexName,
       },
     };
   }
 
-  componentWillReceiveProps(nextProps) {
-    // @TODO: DidUpdate
-    validateNextProps(this.props, nextProps);
-
-    if (this.props.indexName !== nextProps.indexName) {
-      this.aisManager.updateIndex(nextProps.indexName);
-      this.setState(state => ({
-        contextValue: {
-          ...state.contextValue,
-          mainTargetedIndex: nextProps.indexName,
-        },
-      }));
-    }
-
-    if (this.props.refresh !== nextProps.refresh) {
-      if (nextProps.refresh) {
-        this.aisManager.clearCache();
-      }
-    }
+  isUnmounting: boolean = false;
+  aisManager: InstantSearchManager = createInstantSearchManager({
```

Comment:
With the `StrictMode` the manager could be created [more than once](https://reactjs.org/docs/strict-mode.html#detecting-unexpected-side-effects), it sounds incorrect. We'll create the Helper + listeners multiple times, even though they won't collide with each other we spend time for nothing. A  solution would be to move it inside cDU but we've to investigate how feasible it is. Worst case we at least have to check how it behaves inside `StrictMode`  with the current implementation.

Reply:
ah, I wasn't aware constructor also can be called multiple times. cDU happens after render though, that means we render a first time empty for no reason, no?

Reply:
Yes that's why we have to check how it behaves with different solution.

Reply:
If `createInstantSearchManager` does not have any side effect I'm more to keep it here. If we move it to cDU we'll delay the first render each time. Here, at worse, we make some extra work but it is not very significant imho.